src/transformers/models/pop2piano/feature_extraction_pop2piano.py

```diff
@@ -0,0 +1,562 @@
+                feature = audio[start_sample:end_sample]
+                batch.append(feature)
+            return batch
+
+        batch = split_audio(audio)
+        batch = pad_sequence(batch, batch_first=True, padding_value=padding_value)
+
+        return batch, ext_beatstep
+
+    def single_preprocess(
+        self,
+        beatstep,
+        feature_tokens=None,
+        audio=None,
+        n_bars=None,
+    ):
+        """Preprocessing method for a single sequence."""
+
+        if feature_tokens is None and audio is None:
+            raise ValueError("Both `feature_tokens` and `audio` can't be None at the same time!")
+        if feature_tokens is not None:
+            if len(feature_tokens.shape) != 1:
+                raise ValueError(f"Expected `feature_tokens` shape to be (n, ) but found {feature_tokens.shape}")
+        if audio is not None:
+            if len(audio.shape) != 1:
+                raise ValueError(f"Expected `audio` shape to be (n, ) but found {feature_tokens.shape}")
+        n_bars = self.n_bars if n_bars is None else n_bars
+
+        if beatstep[0] > 0.01:
+            warnings.warn(f"Inference Warning : beatstep[0] is not 0 ({beatstep[0]}). all beatstep will be shifted.")
+            beatstep = beatstep - beatstep[0]
+
+        if self.use_mel:
+            batch, ext_beatstep = self.preprocess_mel(
+                audio,
+                beatstep,
+                n_bars=n_bars,
+                padding_value=self.padding_value,
+            )
+        else:
+            raise NotImplementedError("use_mel must be True")
+
+        return batch, ext_beatstep
+
+    def __call__(
+        self,
+        raw_audio: Union[np.ndarray, List[float], List[np.ndarray]],
+        audio_sr: int,
+        steps_per_beat: int = 2,
+        return_tensors: Optional[Union[str, TensorType]] = "pt",
+        **kwargs,
+    ) -> BatchFeature:
+        """
+        Args:
+        Main method to featurize and prepare for the model one sequence. Please note that `Pop2PianoFeatureExtractor`
+        only accepts one raw_audio at a time.
+            raw_audio (`np.ndarray`, `List`):
+                Denotes the raw_audio.
+            audio_sr (`int`):
+                Denotes the Sampling Rate of `raw_audio`.
+            steps_per_beat (`int`, *optional*, defaults to 2):
+                Denotes Steps per beat.
+            return_tensors (`str` or [`~utils.TensorType`], *optional*, defaults to 'pt'):
+                If set, will return tensors instead of list of python integers. Acceptable values are:
+                - `'pt'`: Return PyTorch `torch.Tensor` objects.
+                - `'np'`: Return Numpy `np.ndarray` objects.
+        """
+        warnings.warn(
+            "Pop2PianoFeatureExtractor only takes one raw_audio at a time, if you want to extract features from more than a single audio then you might need to call it multiple times."
+        )
```

Comment:
@ArthurZucker Thanks for your comments! 
I have one more question about this - The feature_extractor outputs variable length tensors (for a audio it outputs `torch.Size([78, 70, 512])` and for another it does `torch.Size([120, 64, 512])`), So if I pad them I need to create another variable to track the shapes when I am about to unpad them in `modeling_pop2piano.py `(if I dont pad them and keep them in list then I wont be able to use `.to("cuda")` since then it will be a list not tensor). And for the model, the model can also take one raw_audio at a time, so should I also use a loop inside `generate`?

Reply:
@ArthurZucker 
After quick reading the code, this implementation is not for training. 

In fact, for training, batch audio inputs of fixed length and corresponding pre-computed features were used (without using feature extractor on-the-fly).

Therefore, currently, in order to inference a single audio with an arbitrary length, single audio is split and batched. Therefore, it is currently not easy to directly implement receiving arbitrary-length batch audio input.